### PR TITLE
Implement Kronecker product of Matrices

### PR DESCRIPTION
**Pull request by raineszm**
This pull request adds a `MatrixExpr` class which represents the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product) of matrices.

#### References to other Issues or PRs

This PR can be seen as a stepping stone to resolving #10771, #10734 , and  #11832 but these commits do not touch the `physics.quantum.tensorproduct` module.

#### Brief description of what is fixed or changed

Add in this PR is the `sympy.matrices.expressions.kronecker` module which exports

* The `kronecker_product()` function, *which attempts to take the Kronecker product of its arguments*
* The `KroneckerProduct` class, *which represents and unevaluated Kronecker product of matrices*
* The `combine_kronecker()` function, *which attempts to combine operations on Kronecker products into Kronecker products of those operations*

The only changes to existing modules in this PR are to add KroneckerProduct to the printing systems and to import the aforementioned functions and class into the `sympy.matrices.expressions` module.

### CODE REVIEW

File: sympy/printing/str.py [hunk 1]
```diff
@@ -311,6 +311,10 @@ def _print_HadamardProduct(self, expr):
         return '.*'.join([self.parenthesize(arg, precedence(expr))
             for arg in expr.args])
 
+    def _print_KroneckerProduct(self, expr):
+        return 'x'.join([self.parenthesize(arg, precedence(expr))
```
Comment (asmeurer) [inline sympy/printing/str.py]: The str printer should be a recreatable form. Pretty printing should go in the pretty printer.

File: sympy/printing/pretty/pretty.py [hunk 2]
```diff
@@ -857,6 +857,15 @@ def _print_HadamardProduct(self, expr):
         return self._print_seq(expr.args, None, None, delim,
                 parenthesize=lambda x: isinstance(x, (MatAdd, MatMul)))
 
+    def _print_KroneckerProduct(self, expr):
+        from sympy import MatAdd, MatMul
+        if self._use_unicode:
+            delim = u' \N{N-ARY CIRCLED TIMES OPERATOR} '
+        else:
+            delim = 'x'
```
Comment (asmeurer) [inline sympy/printing/pretty/pretty.py]: Shouldn't there be a space for the ascii printer too?

**Comment by asmeurer** [comment]
CC @Upabjojr

File: sympy/printing/str.py (hunk 1 above)
Comment (raineszm) [reply sympy/printing/str.py]: Ah, my apologies. I was using the printing of other classes in the `sympy.matrix.expressions` module as a guideline. I'll remove the method in the str printer.

File: sympy/printing/pretty/pretty.py (hunk 2 above)
Comment (raineszm) [reply sympy/printing/pretty/pretty.py]: Yes, there probably should be.

**Comment by jksuom** [comment]
Kronecker product of matrices, which is nowadays usually called tensor product, is also constructed in `sympy.physics.quantum.tensorproduct`.  Could that be made use of here?

**Comment by Upabjojr** [comment]
I believe we should not use the term tensor product for Kronecker product. I would expect the tensor product of two matrices to create a four dimensional array, while the Kronecker product acts like in this PR.

This PR appears to be a step in the right direction.

**Comment by Upabjojr** [comment]
I like this PR. I'm OK to merge this. What do you think?

**Pull request merged**